Add reference semantics to TorchScript classes

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44324**

**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

Commit:
[JIT] Add reference semantics to TorchScript classes

**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -983,6 +983,7 @@ std::shared_ptr<SugaredValue> toSugaredValue(
     return createSimpleEnumValue(obj, m, loc);
   }
 
+  obj = py::module::import("inspect").attr("unwrap")(obj);
```
Overwriting `obj` changes behavior of all following logic, is this intended?

gmagogsfm commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -788,6 +798,11 @@ void initJitScriptBindings(PyObject* module) {
+          .def(
+              "_properties", [](Object& self) { return self.get_properties(); })
+          .def(
+              "equals",
```
should we just consider overriding `__eq__` so that comparison in python is more natural?

gmagogsfm commented on torch/jit/_recursive.py:
```diff
@@ -1,14 +1,17 @@
 import inspect
 import torch
+import pdb
```
accident?

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/api/object.h:
```diff
@@ -102,6 +102,17 @@ struct TORCH_API Object {
+      if (prop.setter) {
+        setter = Method(_ivalue(), prop.setter);
+      }
+      return std::make_tuple(prop.name, Method(_ivalue(), prop.getter), setter);
```
The getter must exist but the setter is optional (for the TorchScript implementation of properties).

SplitInfinity commented on test/jit/test_class_type.py:
```diff
@@ -440,7 +473,7 @@ def __init__(self):
                 return x + self.b
 
-        with self.assertRaisesRegex(RuntimeError, "does not support inheritance"):
+        with self.assertRaisesRegex(TypeError, r"takes 2 positional arguments but 4 were given"):
```
I think this is a consequence of the decorator I added; `Base` is now an instance of `ScriptClassWrapper`, which cannot be inherited from.

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

SplitInfinity commented:
@eellison curious for your thoughts on this, other reviewers seem to be busy.

eellison commented:
@gmagogsfm do you mind reviewing ?  dont know that I have bandwidth for 500+ line PR right now

eellison commented:
@suo is a good person to review this

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

ppwwyyxx commented:
I've tried this PR and it now allows me to run a scripted detectron2 model in C++. Looking forward to it!

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -1035,6 +1127,8 @@ def forward(self, input):
         )
 
     # No-op for modules and functions that are already scripted
```
```suggestion
    # No-op for modules, functions, class instances that are already scripted
```

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -262,6 +290,148 @@ def __getattr__(self, attr):
 
 
 if _enabled:
+    class RecursiveScriptClass(object):
```
What does this mean? Has this comment been resolved offline?

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D24912807).

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D24912807).

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -335,6 +327,102 @@ def __getattr__(self, attr):
+
+        def __getattr__(self, attr):
+            if "_initializing" in self.__dict__ and self.__dict__["_initializing"]:
+                return super(RecursiveScriptClass, self).__getattr__(attr)  # type: ignore
```
```
error: "__getattr__" undefined in superclass  [misc]
```

Commit:
Update on "[JIT] Add reference semantics to TorchScript classes"


**Summary**
This commit adds reference semantics to TorchScript class types;
modifications made to them within TorchScript will be visible in Python.

**Test Plan**
This commit adds a unit test to `TestClassType` that checks that
modifications made to a class type instance passed into TorchScript are
visible in Python after executing the scripted function or module.

**Fixes**
This commit closes #41421, #46944.

Differential Revision: [D24912807](https://our.internmc.facebook.com/intern/diff/D24912807)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D24912807).
